src/core/storage/storage.js

```diff
@@ -90,6 +91,20 @@ export default class GlobalStorage {
     this._callListeners('update', key);
   }
 
+  /**
+   * Sets all of the given entries before calling 'update' event listeners.
+   *
+   * @param {Map} entries the data to set
+   */
+  setEntries (entries) {
```

Comment:
I really like this solution! I'm not sure how this solves the problem of the `query`. Couldn't it still be the case that `query` is the first updated key in the set and then its the first listener called and the remaining aren't called because a search was conducted?

Reply:
I guess searches don't do a page re-load and instead just update the current page

Reply:
nope! since we wait until everything is updated before calling the update listeners

Reply:
So say we call the `StorageKeys.QUERY` update listener and then it starts calling the other listeners with new data (as the StorageKeys.QUERY updated the storage).

Then, because we updated the StorageKeys from the search, we call the other listeners a second time. My worry is that there is a situation here where you wouldn't want this listener to be called twice with the same data (a flash on the page or it causes behavior you wouldn't expect, since the previous listener would be expected to be called with old data).

Am I wrong in this worry?

Reply:
I think your worry is justified but that's not really what this is trying to solve. The problem this pr is trying to solve is trying to make sure that when you want to update multiple storage keys at once, that the listeners don't run until all of them have been updated. The problem you're describing sounds more like "how do we handle people messing with storage inside their listeners?" I think adding setEntries() will help this second problem a little, but you're right it won't solve it. If you have a listener that updates storage you will still have to be careful that it doesn't do something weird, or god forbid infinite loop. But I don't think batching together storage updates, as opposed to calling them one by one, will cause any of the flashes you're describing.

Reply:
leaning towards shelving this for now after talking with christian a little more